bgzf/cache.go

```diff
@@ -125,13 +126,26 @@ func (b *block) Read(p []byte) (int, error) {
 	return n, err
 }
 
+// readToEOF will exhaust the reader or read len(buf) (MaxBlockSize) bytes.
+// If one of those is not true, the err will be non-nil.
+func readToEOF(r io.Reader, buf []byte) (n int, err error) {
+	for err == nil && n < len(buf) {
+		var nn int
+		nn, err = r.Read(buf[n:])
+		n += nn
+	}
+	if err == io.EOF {
+		err = nil
+	} else if err == nil && n != MaxBlockSize {
```

Comment:
In don't understand this. n can only be `0 <= n && n <= MaxBlockSize`. So this branch can never have a nil error according to the loop condition.

What you want to do is

```
switch {
case err == io.EOF:
    return n, nil
case n == MaxBlockSize && err == nil:
    // This is paranoic, but some readers will return
    // quickly when passed a zero-length byte slice.
    var dummy [1]byte
    _, err = r.Read(dummy[:])
    if err == nil {
        return n, io.ErrShortBuffer
    }
    if err = io.EOF {
        err = nil
    }
}
return n, err
```

Add to the comment, readToEOF may consume more bytes from the reader than returned in buf if the length of the available data is greater than MaxBlockSize, or words to that effect.